modules/gapi/test/gapi_async_test.cpp

```diff
@@ -0,0 +1,165 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
+#include "test_precomp.hpp"
+#include "opencv2/gapi/gcomputation_async.hpp"
+#include "opencv2/gapi/gcompiled_async.hpp"
+
+#include <condition_variable>
+
+namespace opencv_test
+{
+struct SumOfSum : ::testing::Test{
+    cv::GComputation sum_of_sum;
+    SumOfSum() : sum_of_sum([]{
+        cv::GMat in;
+        cv::GScalar out = cv::gapi::sum(in + in);
+        return GComputation{in, out};
+    })
+    {}
+};
+
+struct AsyncAPIBasic : SumOfSum {
+    const cv::Size sz{2, 2};
+    cv::Mat in_mat{sz, CV_8U, cv::Scalar(1)};
+    cv::Scalar out;
+
+    cv::GCompiled compile(){
+        return sum_of_sum.compile(descr_of(in_mat));
+    }
+
+    cv::GComputation& computation(){
+        return sum_of_sum;
+    }
+};
+
+template<typename crtp_final_t>
+struct crtp_cast {
+    template<typename crtp_base_t>
+    static crtp_final_t* crtp_cast_(crtp_base_t* this_)
+    {
+        return  static_cast<crtp_final_t*>(this_);
+    }
+};
+
+template<typename crtp_final_t>
+struct CallBack: crtp_cast<crtp_final_t> {
+
+    bool callback_called = false;
+    std::mutex mtx;
+    std::condition_variable cv;
+    std::function<void(std::exception_ptr)> callback(){
+        return [this](std::exception_ptr){
+            mtx.lock();
+            mtx.unlock();
+            cv.notify_one();
+            callback_called = true;
+        };
+    };
+
+    template<typename... Args >
+    void start_async(Args&&... args){
+        this->crtp_cast_(this)->async(callback(), std::forward<Args>(args)...);
+    }
+
+    void wait_for_result()
+    {
+        //FIXME: add a reasonable timeout here ?
+        std::unique_lock<std::mutex> lck{mtx};
+        cv.wait(lck,[&]{return callback_called == true;});
+    }
+};
+
+template<typename crtp_final_t>
+struct Future: crtp_cast<crtp_final_t> {
+    std::future<void> f;
+
+    template<typename... Args >
+    void start_async(Args&&... args){
+        f = this->crtp_cast_(this)->async(std::forward<Args>(args)...);
+    }
+
+    void wait_for_result()
+    {
+        //FIXME: add a reasonable timeout here ?
```

Comment:
Can you add timeout argument to the function, with default value (1 min?)

Reply:
After some thinking it out - it was a bad idea to add timeout - as there is no way to cancel the work gracefully , so timeout simply meaningless/confusing here, so TODO is removed